src/librustc_middle/ty/mod.rs

```diff
@@ -1615,12 +1615,33 @@ pub struct WithOptConstParam<T> {
 
 impl<T> WithOptConstParam<T> {
     /// Creates a new `WithOptConstParam` setting `const_param_did` to `None`.
+    #[inline(always)]
     pub fn unknown(did: T) -> WithOptConstParam<T> {
         WithOptConstParam { did, const_param_did: None }
     }
 }
 
 impl WithOptConstParam<LocalDefId> {
+    /// Returns `Some((did, param_did))` if `def_id` is a const argument,
+    /// `None` otherwise.
+    #[inline(always)]
+    pub fn try_fetch(did: LocalDefId, tcx: TyCtxt<'_>) -> Option<(LocalDefId, DefId)> {
```

Comment:
I think it's clearer if you start with `unknown` and then use the other method. Bonus: the other case likely takes the `unknown` anyway, so it would only be created once.

Reply:
I don't think so. The other method creates a `WithOptConstParam`, requiring us to write `query_name((def.did, def.const_param_did.unwrap()))` which seems really noisy.

Reply:
We might also change the return value of `try_upgrade` to `Option<(LocalDefId, DefId)>` but that makes the other use case less clear.

The problem at its core is that we have two different kinds of queries.
- split queries with `query_name` and `query_name_of_const_arg` where `try_fetch` is used at the start of `query_name`
- shared queries taking `ty::WithOptConstParam`, these call `try_upgrade` at their start and call itself if that returns some.

Reply:
We may try and split the remaining shared queries at which point `try_upgrade` won't be used anymore.